compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -390,16 +426,19 @@ object Objects:
      *
      * @return the environment and value for `this` owned by the given method.
      */
-    def resolveEnv(meth: Symbol, thisV: Value, env: Data)(using Context): Option[(Value, Data)] = log("Resolving env for " + meth.show + ", this = " + thisV.show + ", env = " + env.show, printer) {
+    def resolveEnv(meth: Symbol, thisV: ThisValue, env: Data)(using Context): Option[(ThisValue, Data)] = log("Resolving env for " + meth.show + ", this = " + thisV.show + ", env = " + env.show, printer) {
       env match
       case localEnv: LocalEnv =>
         if localEnv.meth == meth then Some(thisV -> env)
         else resolveEnv(meth, thisV, localEnv.outer)
       case NoEnv =>
-        // TODO: handle RefSet
         thisV match
         case ref: OfClass =>
-          resolveEnv(meth, ref.outer, ref.env)
+          ref.outer match
+          case outer : ThisValue =>
+            resolveEnv(meth, outer, ref.env)
+          case _ => // This is the case for top-level classes
+            None
```

Comment:
If I remember correctly, outer can be RefSet. We might want to do something here. We can fix this in another PR. Please correct the comment and add back the TODO.

The original code `ref.outer` looks dubious: Why blindly go to the outer of the most specific class? It should still be correct but for a different reason: If the parent class is local, then the local environment must be accessible from the subclass.